### DISCUSSION TITLE
Can I force all listeners to immediately rerender after calling queryCache.setQueryData()?

### DISCUSSION THREAD

**Discussion by justincy** [TanStack/query#758] [General] [answered]
I'm using react-query to keep track of the user's account. When they login, I call `queryCache.setQueryData('account', account)` but that doesn't immediately force all account query listeners to rerender. How can I make that happen?

**Comment by tannerlinsley** [comment, marked as answer]
When you call `queryCache.setQueryData('account', ...)`, any queries that use the `'account'` key should and will be updated. If they don't, please create a codesandbox illustrating the issue so we can better debug it.

**Reply by justincy** [reply]
Thanks. It turns out my problem was outstanding queries that resolved after calling `setQueryData()`. I didn't expect - nor do I want - those extra queries. Calling `cancelQueries()` works, mostly. It leads to a ton of rerenders and UI thrashing, but it does eventually resolve correctly. I'd prefer to just prevent that outstanding query. Let me explain my situation in more detail.

1. I'm using react-query to cache the user account.
2. When the user logs out, I forward them to a page that shows a spinner while issuing a request to the server to clear the HTTP-only session cookies.
3. When rendering, the page has in it's hierarchy a `useQuery('account')` call. That account query is refetching data, unnecessarily. Not just because I know I don't need it while logging out, but also because I'm trying to set config on the query such that it never refreshes. But I can't seem to find the right set of options. How do I do that?

I currently have this config:

```js
{
  refetchInterval: false,
  refetchOnWindowFocus: false,
  refetchIntervalInBackground: false,
  retry: false
}
```

That prevents many unnecessary fetches, but not all, including this unnecessary fetch during logout. I've tried adding `staleTime: Infinity` and `cacheTime: Infinity` but then updates via `setQueryData()` are ignored.

So how can I prevent a query from going stale and refetching while also being able to update it via `setQueryData()`?

**Reply by cherniavskii** [reply]
So you want to fetch account query *only* if user is logged in?
Sounds like a great use case for dependent query (assuming that you know if your user is logged in):

```js
useQuery('account', queryFn, {
  enabled: isUserLoggedIn,
})
```

**Reply by justincy** [reply]
> So you want to fetch account query only if user is logged in?

Sort of. I want to fetch it only _once_ when the user is logged in, but I want to reserve the ability to update it via `setQueryData()`.

I've been thinking about using the `enabled` option, but in a different manner: set it to false after the first time `onSuccess` is called.

**Reply by tannerlinsley** [reply]
You can set enabled to false permanently and just use the queryInfo.refetch method to manually fetch when you want.

**Reply by justincy** [reply]
I think I'm just going to switch back to using context for this. It makes more sense for the behavior I want.

**Reply by tannerlinsley** [reply]
I think that’s a good idea. FWIW, I use a custom context for auth myself.

**Comment by mikhailgagarin** [comment]
@tannerlinsley I have a similar issue. Here is a little representation of it:
https://codesandbox.io/s/gallant-dawn-rsscv
So initially I have a default object and after the user sign-in there should be placed the user info, but it works only after the 2nd click on the "Sign in" button.

**Reply by TkDodo** [reply]
I updated your example to v3.6.0 and it works correctly there: https://codesandbox.io/s/hungry-oskar-rrmwo?file=/src/App.js

**Reply by mikhailgagarin** [reply]
@TkDodo Thank you! But unfortunately for now I'm sticking with v2. If there is a way to fix my code without upgrading a version of the package it'd be great.
